### ISSUE TITLE
Roadmap v1.0.0

### ISSUE THREAD

**Issue by S1SYPHOS** [rosell-dk/webp-convert#8]
Hey there,
seems to me that we shouldn't include `vendor` - it's considered [bad practice](https://getcomposer.org/doc/faqs/should-i-commit-the-dependencies-in-my-vendor-directory.md) **and** we simply don't need it. Last, `webp-convert.php` seems redundant to me - please enlighten me!

Also, I started refactoring code on branch `patch-1`, containing API-breaking changes such as
- [x] proper namespacing
- [x] camel-casing variables & function names
- [x] extracting functions (mostly from `convert()`)
- [x] PSR-2 code compliance
- [x] flat file structure
- [x] adding tests via `PHPUnit` ..
- [x] .. and automating them with Travis CI

:crossed_fingers: 

Cheers!

**Comment by S1SYPHOS** [comment]
Also, I started applying PSR2 coding rules via PHP CodeSniffer, coming right up which also requires 

- [ ] proper namespacing
- [ ] camel-case function names

**Comment by rosell-dk** [comment]
I have also felt an urge to get rid of webp-convert.php. What held me back was that I need the functionality in all three projects where I use WebPConvert. So getting rid of it here will force me to add it in those three projects. Also redundant...

But ok, lets get rid of it in this repository. It adds little value to this projects. Or, perhaps, it even subtracts value by adding confusion and being messy.

**Comment by rosell-dk** [comment]
Oh, sorry - the danger of _git add -A_. The vendor folder was not supposed to be added. Good thing that you noticed

**Comment by rosell-dk** [comment]
With webp-convert.php gone, there is no need for WebPConvertPathHelper.php either. I'm removing it.
Also, I'm adding a .gitignore to avoid vendor being accidently added again

**Comment by S1SYPHOS** [comment]
I‘d like to comply with PSR2, so is changing function names to camelcase okay with you? I guess you will have to change stuf in your other Plugins?

**Comment by S1SYPHOS** [comment]
And namespacing? Maybe even separating includes & WebPConvert class?

**Comment by rosell-dk** [comment]
PSR2 compliance is ok with me.

**Comment by rosell-dk** [comment]
Yes, it is going to be an API-breaking feature. It's ok. We have other incoming API-breaking features... (#14)

**Comment by rosell-dk** [comment]
I have no experience in PSR2, so I hope that feature is on you ?

**Comment by rosell-dk** [comment]
I suggest we implement and test PSR2 and #14 one at the time

**Comment by rosell-dk** [comment]
I guess I should be involved in the PSR2 compliance after all. So I don't start breaking the compliance as soon as we achieved it.

**Comment by S1SYPHOS** [comment]
Not really a feature, mostly syntax, but I‘m on it! Just wanted to make sure that you‘re okay with it :dancer:

**Comment by rosell-dk** [comment]
Yeah, atom here too :) - And I actually also searched and found the PHP CS fixer script, and the atom plugin for it. It however doesn't work for me out of the box, so I will have some configuration to do

**Comment by S1SYPHOS** [comment]
You have to install php_codesniffer and insert its path in plugin settings :cat:

**Comment by rosell-dk** [comment]
thanks. I got it working now

**Comment by S1SYPHOS** [comment]
I‘m working on the branch patch, as soon as I‘m satisfied, I‘ll open a PR for review!

**Comment by rosell-dk** [comment]
ok, great :)

**Comment by S1SYPHOS** [comment]
PR up and running! The remaining points on the agenda abhöbe will be fulfilled after the merge, since they‘re non-critical: documentation, travis & tests.

I‘ll update the README soon, and that‘s it Thema for upcoming release :champagne:

**Comment by S1SYPHOS** [comment]
Hey there,
I just created a milestone for the upcoming release.

**Proposal:**
When we are happy with the changes on `patch-1` branch, I suggest deleting the current release and go for v1.0.0. Also, the Packagist auto-update feature should be revisited so all future changes get included automatically.

How does that sound to you?

**Comment by rosell-dk** [comment]
Why remove current release?

**Comment by S1SYPHOS** [comment]
It seems somewhat weird, jumping from 0.1.0 to 1.0.0 - I'd rather go for a fresh start, but your choice. In the meantime, my solution for #17 works perfectly, I'll push it onto `patch-1`, closing the issue!

**Comment by rosell-dk** [comment]
We are using semantic versioning, so esthetic considerations can be put aside. Bug-fix-only releases triggers the last number. New features, no API changes triggers middle number. API changes triggers first number. This versioning is clever, because it can be used i.e when setting up composer.json. Users are relying on us keeping to the rules. If they configure updating not to allow API-changing updates, they are relying on that updates won't break their code.

I think we should keep the version. I still have stuff running on 0.1.0. It is practical for me to easily get that release. And generally, I guess it is bad practice to remove old versions

**Comment by S1SYPHOS** [comment]
Alright, sounds Good!

**Comment by daybugging** [comment]
Great stuff, hopefully this gets released soon. As I wrote on the other issue, I'm already using it anyway 😄

**Comment by rosell-dk** [comment]
Yes, good stuff. And codos to S1SYPHOS for lifting this library into maturity. Just a little bit of further patience is needed, to assure that we get a stable API, which provides what will be needed and does not need changing in near future. But we are on it!

**Comment by S1SYPHOS** [comment]
Glad you like it so far, bear with me!

**Comment by rosell-dk** [comment]
I have:
- merged PATCH-1
- implemented the requirements I wrote in the comments in PATCH-1
- made various bug fixes
- done some testing

I uncommented the tests of the functions that are now private. And I also uncommented the unit tests of the converters. It is not a bug, that a converter fails to convert. The result is that there are no tests left! It would be possible to add unit tests of the convert methods for testing that they fail in various cases (ie, when target file is known not to be found). I don't see much value in such tests, though. I think we can remove unit tests from this road map.

I don't see any tasks left. #16 isn't critical. It can come in 1.0.1. 

But we should fine-read the README, skim the code again, and do a bit more testing, just to make sure that there are no obvious bugs.

**Comment by rosell-dk** [comment]
What we *could* do regarding unit tests, is to test that they either produce an Exception OR an image

**Comment by S1SYPHOS** [comment]
So writing all those tests was completely in vain? :disappointed:

**Comment by rosell-dk** [comment]
Perhaps.
But it is great that you got phpunit and travis ci set up.

**Comment by rosell-dk** [comment]
I have added unit tests that tests conversion and various errors.

**Comment by rosell-dk** [comment]
Travis CI has been activated.
But we have a build error in hhvm.

**Comment by rosell-dk** [comment]
I plan to add some unit tests for the converters as well.

**Comment by S1SYPHOS** [comment]
Alright then, but since v1.0.0 is released and this concern was moved to another issue, I'll close this again!

**Comment by rosell-dk** [comment]
But v1.0.0 is not released yet - fortunately! I say fortunately, because I'm not sure we made the right decision about convert throwing exceptions.
Please don't release it until we both say GO. Opening again :)

**Comment by rosell-dk** [comment]
See #24

**Comment by S1SYPHOS** [comment]
Mea culpa, I thought merging my changes meant releasing.

**Comment by S1SYPHOS** [comment]
Just want to let you know that I started an independent fork where I want to test different ideas without accidentally merging them into `webp-convert`. It'll be heavily under construction, and therefore may only be a temporary sidetrack - however, this doesn't affect my work for the main library :champagne:

**Comment by rosell-dk** [comment]
Ok!

**Comment by rosell-dk** [comment]
Note that I'm working quite a lot on this project now, so merging with git may not be painless

**Comment by S1SYPHOS** [comment]
I got that and will check out your progress whenever I got the time. If in need, just ping me - watching this repository spams my feed too hard :space_invader:

**Comment by rosell-dk** [comment]
I think we should start out with a 1.0.0-beta. I suppose we are ready for that now. Or?

**Comment by rosell-dk** [comment]
Its on packagist now. And I fixed the packagist hook, so next time, it will update automatically

**Comment by S1SYPHOS** [comment]
Hey again,
[`php-curl-class`](https://github.com/php-curl-class/php-curl-class) satisfies all our cURL needs, even key validation is super-easy for `Ewww` (and future cURL-based providers).

Example:
```text
composer require php-curl-class/php-curl-class
```
```php
<?php
    // Throws an exception if the provided API key is invalid
    public static function isValidKey($key = insertKeyFromOptionsArrayHere)
    {
        try {
            $curl = new \Curl\Curl();
            $result = $curl->post('https://optimize.exactlywww.com/verify/', [
                'api_key' => $key
            ]);
            if ($curl->error) {
                throw new \Exception($curl->errorMessage . ' - ' . $curl->errorCode);
            }
        } catch (\Exception $e) {
            return false; // TODO: Throw $e so error may be inspected later
        }
        /*
         * There are three possible responses:
         * 'great' = verification successful
         * 'exceeded' = indicates a valid key with no remaining image credits
         * '' = an empty response indicates that the key is not valid
        */
        $success = (bool) preg_match('/great/', $result);
        if (!$success && preg_match('/exceeded/', $result)) {
            throw new \Exception('API key is valid, but has no remaining image credits.');
        }
        return $success;
    }
```
```php
<?php

// inside its convert() function:
        try {
            $this->checkRequirements();
            // Checking if provided key is valid
            $this->isValidKey();
            // Initializing cURL, setting response headers & requesting image conversion
            $curl = new \Curl\Curl();
            $curl->setHeader('User-Agent: WebPConvert', 'Accept: image/*');
            $result = $curl->post('https://optimize.exactlywww.com/v2/', [
                'api_key' => WEBPCONVERT_EWWW_KEY,
                'webp' => '1',
                'file' => curl_file_create($this->source),
                'quality' => $this->quality,
                'metadata' => ($this->strip ? '0' : '1')
            ]);
            if ($curl->error) {
                throw new \Exception($curl->errorMessage . ' - ' . $curl->errorCode);
            }
        } catch (\Exception $e) {
            return false; // TODO: `throw` custom \Exception $e & handle it smoothly on top-level.
        }
        $success = file_put_contents($this->destination, $result);
        if (!$success) {
            return false;
        }
        return true; 
```

**Comment by rosell-dk** [comment]
The Curl-class library seems to be just wrapping curl functionality. So basically, it is syntactic sugar. I understand that someone felt the urge to provide this sugar, though, because curl doesn't strike me as having very friendly API. But I'm more comfortably with using curl directly. Our code only gets a bit simpler with a switch, but at a price of adding a complex library. We risk that the library changes API, which means our code may stop working. It is not a good sign that the API is is on major version 8.

**Comment by S1SYPHOS** [comment]
Good thinking, glad I ran this by you, as always.
It certainly helps reducing code complexity in our scripts, additionally 
all my efforts to get Optimus.io working with php-curl-class failed - 
whereas using the provided cURL example worked out-of-the-box.

**Comment by rosell-dk** [comment]
I think #38 is a 1.0.0 blocker, as I don't want WebPConvert to put pressure on the Ewww library. But we could take ewww out of the library and add it again in 1.1.0, if we want to release 1.0.0 soon

**Comment by rosell-dk** [comment]
The Ewww cloud service is protected now. But actually, I'm in no hurry with the 1.0.0 release. I'm rewriting WebPOnDemand, WebPRealizer and WebPExpress to use the new library, and I'd like to see how it all goes. They all need some functionality that have been removed in WebPConvert.

I have created this new library, in which some of the common functionality I need now resides:
https://github.com/rosell-dk/webp-convert-and-serve
